gwu-libraries/sfm-twitter-harvester
# T1104 v2 filter stream

Resolves gwu-libraries/sfm-ui#1104

### Contains
- Upgrade to twarc 2.13, twarc-csv 0.7
- Added new Dockerfile for Twitter stream exporter (v2 exports)
- twitter_harvester.py
  - Added error handling/user-friendly error messages for v.2 API error codes
  - Added support for v.2 streaming rules: 
    - New seeds are treated as additions to the rules
    - Any rules registered not present in the current active seeds are deleted via API
    - Error handlings for rules cap exceeded (cap varies by API access level)
- twitter_rest_warc_iter.py
  - Fixed issue whereby filtering by Twitter User ID wasn't working
- twitter_stream_warc_iter.py
  - Support for iterating over Twitter v.2 JSON (from Filtered Stream API)
  - Note that filtering results by Twitter User ID is not currently supported for this harvest type
- twitter_stream_exporter.py
  - Added logic to delegate to the appropriate version (instance) of the stream exporter, v.1 or v. 2

### Testing
- Harvesting
  - Without Academic Search credentials, create a collection of this type and turn on harvesting. Verify that an appropriate error message is displayed and that the harvest is turned off
  - Verify (from UI and logs) that a Filtered Stream harvest runs and collects Tweets
  - Verify that stopping and starting a Filtered Stream collection from the UI has the intended effect
  - Add more streaming rules than allowed by your API access level -- verify that the appropriate warning appears in the SFM UI
  - Remove streaming rules via the UI and verify that the rules are updated prior to the next harvest (can use the twarc2 command-line utility to check this)
- Exporting
  - Perform exports (all types) from a Filtered Stream collection
  - Verify that the number of Tweets recorded in the UI matches the number in the export
  - Verify that streaming rules appear in the exported results
  - Verify that limiting an export by harvest and item date works as expected


### Testing
